### PR TITLE
[1.11.2] Add MerchantTradeEvent

### PR DESCRIPTION
**Pull request by CrafterKina**
This pr adds MerchantTradeEvent. 

Usecase
---
* Give an achievement
* Let the villagers refuse the trade
* Advanced trading system

etc...

### CODE REVIEW

File: src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        super(merchant.getCustomer());
+        this.merchant = merchant;
+        this.trade = trade;
+        this.left = left;
+        this.right = right;
+    }
+
+    @Nonnull
+    public IMerchant getMerchant(){
```
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java]: formatting

File: src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+ * {@link #right} contains the itemstack in the right side.<br>
+ * <br>
+ * All children of this event are fired on the {@link MinecraftForge#EVENT_BUS}.
+ **/
+public class MerchantTradeEvent extends PlayerEvent
+{
+    private final IMerchant merchant;
+    private final MerchantRecipe trade;
+    protected ItemStack left;
+    protected ItemStack right;
```
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java]: Keep two `protected final` fields of original left and right contents.

**Review by liach — CHANGES_REQUESTED** [review]

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
Should the 2 subclasses be named `Pre` and `Post`?

File: src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+     * This event is not {@link Cancelable}.<br>
+     * <br>
+     * This event has a result. {@link HasResult}<br>
+     * {@link Result#ALLOW} The result of trade appears.<br>
+     * {@link Result#DENY} The result of trade does not appear.<br>
+     * <br>
+     * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+     **/
+    @HasResult
+    public static class CanTrade extends MerchantTradeEvent
```
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java]: I would name this `SetupOffer`

File: src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+     * <br>
+     * This event is {@link Cancelable}.<br>
+     * If this event is canceled, the itemstack being about to be sold by the player is not shrunk.<br>
+     * <br>
+     * This event does not have a result. {@link HasResult}<br>
+     * <br>
+     * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+     **/
+    @Cancelable
+    public static class Done extends MerchantTradeEvent
```
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java]: I would name this one `Transact`, specifying that this is fired after the recipe is already created and when the player do the actual transaction.

**Review by liach — APPROVED** [review]

**Comment by mezz** [comment]
It looks like `@Mod.EventBusSubscriber` does not seem to work for the test mods, probably because it's in the same folder as other mods. Can you make it subscribe the old way until we can fix this issue?

Also, please make the test mod disabled by default, and add a comment explaining what `onMerchantTradeDone` is supposed to do so that it can be checked more easily in the future.

**Comment by mezz** [comment]
After some debugging, I found that `@Mod.EventBusSubscriber` works correctly but the methods marked `@SubscribeEvent` need to be `static` to be subscribed.

**Comment by CrafterKina** [comment]
Fixed

**Comment by mezz** [comment]
It works correctly, but the way the events are set up make the interaction strange.
Instead of setting what you will get ahead of time, the event changes it as you try to interact.
I think that doing it this way is not as intuitive for players as it could be.

Event says you can't trade if experience is < 1,  but the player has no idea why the trade doesn't work:


Event says >20 levels makes the cost only 1, but there's no way to see that until you try paying the full price:

**Comment by CrafterKina** [comment]
Should I add a hook to ``isRecipeDisabled`` and set events non-Cancelable?

**Comment by mezz** [comment]
I think that makes sense.
If the price is changed or the trading is disabled, it should show that in the gui.

**Comment by liach** [comment]
@CrafterKina @mezz Any updates?

**Comment by mezz** [comment]
I am closing this PR for now because you have not responded in a long time.
It seems to be nearly complete. Please update and reopen it, or someone else update it and make a new PR.

**Pull request closed without merging**